safekeeper/src/wal_storage.rs

```diff
@@ -445,19 +469,17 @@ impl Storage for PhysicalStorage {
             self.fdatasync_file(&unflushed_file).await?;
             self.file = Some(unflushed_file);
         } else {
-            // We have unflushed data (write_lsn != flush_lsn), but no file.
-            // This should only happen if last file was fully written and flushed,
-            // but haven't updated flush_lsn yet.
-            if self.write_lsn.segment_offset(self.wal_seg_size) != 0 {
-                bail!(
-                    "unexpected unflushed data with no open file, write_lsn={}, flush_lsn={}",
-                    self.write_lsn,
-                    self.flush_record_lsn
-                );
-            }
+            // We have unflushed data (write_lsn != flush_lsn), but no file. This
+            // shouldn't happen, since the segment is flushed on close.
```

Comment:
This still might happen because write_in_segment temporarily takes the file; if fsync/rename fails it is not installed back. So either assert should be removed completely (and file reopened) or write_in_segment shouldn't take the file out.

This is not directly related to this PR and shouldn't be problem in practice because on error reconnection would do truncate_wal establishing self.flush_record_lsn == self.write_record_lsn, so I'm ok with leaving this out.

Reply:
Ok, I'll have another look at this in a follow-up, since I don't believe this changes the current behavior.